### PR TITLE
[#362] Remove scripts, script addresses, and script witnesses

### LINKED ISSUE #362: Remove Scripts, Script Addresses, and Script Witnesses
## Background

In cardano-sl we used `Script`s from `plutus-prototype`, but these never got used on the chain. They will not be used in the OBFT era, so we should remove them to simplify the code.

## Proposed Solution

Remove `ScriptASD` and `ATScript` from `AddrSpendingData` and `AddrType`, respectively. Remove `ScriptWitness` from `TxInWitness`. Remove all remaining references to `Script`s in the code.

### PR DESCRIPTION
**Pull request by ruhatch**
Closes #362

### CODE REVIEW

File: src/Cardano/Chain/Common/AddrSpendingData.hs [hunk 1]
```diff
@@ -12,131 +14,77 @@ where
...
-    = PubKeyASD !PublicKey
-    -- ^ Funds can be spent by revealing a 'PublicKey' and providing a
-    -- valid signature.
-    | ScriptASD !Script
-    -- ^ Funds can be spent by revealing a 'Script' and providing a
-    -- redeemer 'Script'.
-    | RedeemASD !RedeemPublicKey
-    -- ^ Funds can be spent by revealing a 'RedeemPublicKey' and providing a
-    -- valid signature.
-    | UnknownASD !Word8 !ByteString
```
Comment (intricate) [inline src/Cardano/Chain/Common/AddrSpendingData.hs:36]: Is this sort of forwards compatibility no longer important?

Comment (intricate) [reply src/Cardano/Chain/Common/AddrSpendingData.hs:36]: Nevermind my comment, just saw #344.

**Review by mhuesch — APPROVED** [review]
✂️ ✂️ ✂️ 

LGTM

**Review by intricate — APPROVED** [review]

File: src/Cardano/Chain/Common/AddrSpendingData.hs [hunk 2]
```diff
@@ -12,131 +14,77 @@ where
...
-If tag is greater than 2, the data is decoded as a plain 'ByteString'.
-
-This lets us have backwards compatibility. For instance, if a newer
-version of CSL adds a new type of spending data with tag 3, then older
-versions would deserialize it as follows:
-
-    UnknownASD 3 <some bytes>
--}
-
--- Helper function to avoid writing `:: Word8`.
```
Comment (dnadales) [inline src/Cardano/Chain/Common/AddrSpendingData.hs:71]: :thinking:

File: src/Cardano/Chain/Common/AddrSpendingData.hs [hunk 3]
```diff
@@ -12,131 +14,77 @@ where
...
+  encode = \case
+    PubKeyASD pk -> encodeListLen 2 <> encode (0 :: Word8) <> encode pk
+    RedeemASD redeemPK ->
+      encodeListLen 2 <> encode (2 :: Word8) <> encode redeemPK
+
+  decode = do
+    enforceSize "AddrSpendingData" 2
+    decode @Word8 >>= \case
+      0   -> PubKeyASD <$> decode
+      2   -> RedeemASD <$> decode
```
Comment (dnadales) [inline src/Cardano/Chain/Common/AddrSpendingData.hs:53]: I don't know whether it'd make sense to leave a comment explaining why this gap.

Comment (ruhatch) [reply src/Cardano/Chain/Common/AddrSpendingData.hs:53]: :+1:

File: src/Cardano/Chain/Update/ProtocolParameterUpdate.hs [hunk 4]
```diff
@@ -20,15 +20,15 @@ import Formatting (Format, bprint, build, bytes, later, shortest)
...
+import Cardano.Chain.Common (LovelacePortion, TxFeePolicy)
 import Cardano.Chain.Slotting (EpochIndex, FlatSlotId(..))
 import Cardano.Chain.Update.ProtocolParameters (ProtocolParameters(..))
 import Cardano.Chain.Update.SoftforkRule (SoftforkRule)
 
 
 -- | Data which represents modifications of block (aka protocol) version
 data ProtocolParameterUpdate = ProtocolParameterUpdate
-  { ppuScriptVersion     :: !(Maybe ScriptVersion)
+  { ppuScriptVersion     :: !(Maybe Word16)
```
Comment (dnadales) [inline src/Cardano/Chain/Update/ProtocolParameterUpdate.hs:31]: Just for my understanding, why did we replace `ScriptVersion` to `Word16`?

File: test/Test/Cardano/Chain/Update/Example.hs [hunk 5]
```diff
@@ -101,7 +96,7 @@ exampleProtocolParameters = ProtocolParameters
 
 exampleProtocolParameterUpdate :: ProtocolParameterUpdate
 exampleProtocolParameterUpdate = ProtocolParameterUpdate
-  (Just (999 :: ScriptVersion))
+  (Just (999 :: Word16))
```
Comment (dnadales) [inline test/Test/Cardano/Chain/Update/Example.hs:99]: Funny, I thought GHC should be able to infer the type of `999`...

**Review by dnadales — APPROVED** [review]
Zen!

:wave:

File: src/Cardano/Chain/Update/ProtocolParameterUpdate.hs (hunk 4 above)
Comment (ruhatch) [reply src/Cardano/Chain/Update/ProtocolParameterUpdate.hs:31]: It was just a type synonym and I didn't want to keep the `Script` file around that defined it, but did want to keep this parameter around for backwards compatibility.

Comment (dnadales) [reply src/Cardano/Chain/Update/ProtocolParameterUpdate.hs:31]: Oh, that makes sense! Thanks.

**Pull request merged**